### PR TITLE
Unit tests for S3 + parquet

### PR DESCRIPTION
**Pull request by malhotrashivam**
Related to #5064 

This PR should be merged after #5493

### CODE REVIEW

File: extensions/s3/build.gradle [hunk 1]
```diff
@@ -25,14 +25,14 @@ dependencies {
     Classpaths.inheritAutoService(project)
     Classpaths.inheritImmutables(project)
 
-    Classpaths.inheritJUnitPlatform(project)
+    Classpaths.inheritJUnitClassic(project, 'testImplementation')
```
Comment (malhotrashivam) [inline extensions/s3/build.gradle:28]: I had to change from Junit5 to Junit4 because I wanted to use the following to set up a query context for making `Table` objects in `S3ParquetTests` class
```
    @Rule
    public final EngineCleanup framework = new EngineCleanup();
```
@devinrsmith can you let me know if this is okay or there is a better way to do this?

File: extensions/s3/src/test/java/io/deephaven/extensions/s3/S3SeekableChannelTestBase.java [hunk 2]
```diff
@@ -96,17 +101,31 @@ public int read() {
...
+        try (final Stream<Path> paths = Files.walk(directoryPath)) {
+            paths.filter(Files::isRegularFile)
+                    .forEach(file -> {
+                        final String key = directoryName + "/" + directoryPath.relativize(file);
+                        putObject(key, RequestBody.fromFile(file.toFile()));
+                    });
+        } catch (final IOException e) {
+            throw new UncheckedDeephavenException("Failed to walk directory: " + directory, e);
+        }
+    }
```
Comment (devinrsmith) [inline extensions/s3/src/test/java/io/deephaven/extensions/s3/S3SeekableChannelTestBase.java:125]: I've added a more general way to upload directories to S3 using the S3Transfer manager; see https://github.com/deephaven/deephaven-core/pull/5442.

File: extensions/s3/build.gradle (hunk 1 above)
Comment (devinrsmith) [reply extensions/s3/build.gradle:28]: I'm really trying to move us forward to JUnit 5, especially with new code. If absolutely necessary, you can use JUnit 5 runner against JUnit 4 tests; I believe there may be an example or two of that we use elsewhere.

That said, I _don't_ think we should be moving engine tests into the S3 layer - I think the strategy of bringing in S3 testing to the engine (/parquet) is the better approach. I think this should be doable after #5442.

File: extensions/parquet/table/build.gradle [hunk 3]
```diff
@@ -52,7 +52,6 @@ dependencies {
     Classpaths.inheritSlf4j(project, 'slf4j-simple', 'testRuntimeOnly')
 
     runtimeOnly project(':extensions-trackedfile')
-    testImplementation project(':extensions-s3')
```
Comment (devinrsmith) [inline extensions/parquet/table/build.gradle:55]: This can be changed to 

```gradle
testImplementation TestTools.projectDependency(project, 'extensions-s3')
```

to pull the testlib into parquet tests.

**Review by devinrsmith — COMMENTED** [review]
I don't want to invert the dependency here unless we absolutely need to.

**Review by chipkent — APPROVED** [review]
Python LGTM

File: extensions/parquet/table/src/test/java/io/deephaven/parquet/table/ParquetS3SimpleTestBase.java [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+import java.util.concurrent.TimeoutException;
+
+import static io.deephaven.engine.testutil.TstUtils.assertTableEquals;
+import static io.deephaven.engine.util.TableTools.merge;
+import static io.deephaven.parquet.table.ParquetTools.writeKeyValuePartitionedTable;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+@Category(OutOfBandTest.class)
+abstract class ParquetS3SimpleTestBase extends S3SeekableChannelTestSetup {
```
Comment (devinrsmith) [inline extensions/parquet/table/src/test/java/io/deephaven/parquet/table/ParquetS3SimpleTestBase.java:40]: I called the other tests "simple" sort of to hint that they were only looking superficially at some bytes. I think this would be better named "S3ParquetTestBase". Similar w/ the derived classes.

File: extensions/parquet/table/src/test/java/io/deephaven/parquet/table/ParquetS3SimpleTestBase.java [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+
+    // The following tests are disabled by default, as they are verifying against a remote system
+    private static final boolean ENABLE_REMOTE_S3_TESTING = false;
+
+    private static final File rootDir = new File(ParquetS3SimpleTestBase.class.getName() + "_root");
```
Comment (devinrsmith) [inline extensions/parquet/table/src/test/java/io/deephaven/parquet/table/ParquetS3SimpleTestBase.java:45]: We should probably use something like:

```java
    @Rule
    public TemporaryFolder folder = new TemporaryFolder();
```

File: extensions/parquet/table/src/test/java/io/deephaven/parquet/table/ParquetS3SimpleMinIOTest.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+import software.amazon.awssdk.services.s3.S3AsyncClient;
+
+import static org.junit.Assert.assertFalse;
+
+public class ParquetS3SimpleMinIOTest extends ParquetS3SimpleTestBase {
+
+    @BeforeClass
+    public static void initContainer() {
+        // TODO(deephaven-core#5116): MinIO testcontainers does not work on OS X
+        assertFalse("OSUtil.runningMacOS()", OSUtil.runningMacOS());
```
Comment (devinrsmith) [inline extensions/parquet/table/src/test/java/io/deephaven/parquet/table/ParquetS3SimpleMinIOTest.java:20]: This needs to be an _assumption_ not an _assertion_. `org.junit.Assume`

File: extensions/parquet/table/src/test/java/io/deephaven/parquet/table/ParquetS3SimpleTestBase.java [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+import static io.deephaven.engine.util.TableTools.merge;
+import static io.deephaven.parquet.table.ParquetTools.writeKeyValuePartitionedTable;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+@Category(OutOfBandTest.class)
+abstract class ParquetS3SimpleTestBase extends S3SeekableChannelTestSetup {
+
+    // The following tests are disabled by default, as they are verifying against a remote system
+    private static final boolean ENABLE_REMOTE_S3_TESTING = false;
```
Comment (devinrsmith) [inline extensions/parquet/table/src/test/java/io/deephaven/parquet/table/ParquetS3SimpleTestBase.java:43]: We should not intermingle testcontainer tests and testing against a real remote. As such, all tests that reference `ENABLE_REMOTE_S3_TESTING` should likely be in a dedicated testing class.

File: Base/src/main/java/io/deephaven/base/FileUtils.java [hunk 8]
```diff
@@ -282,21 +282,30 @@ public static URI convertToURI(final String source, final boolean isDirectory) {
...
+                // Convert to a "file" URI
+                return convertToURI(new File(source), isDirectory);
+            }
+            String path = uri.getPath();
+            boolean isUpdated = false;
+            // Directory URIs should end with a slash
+            if (isDirectory && path.charAt(path.length() - 1) != URI_SEPARATOR_CHAR) {
+                path = path + URI_SEPARATOR_CHAR;
+                isUpdated = true;
+            }
```
Comment (devinrsmith) [inline Base/src/main/java/io/deephaven/base/FileUtils.java:295]: We should probably add some unit tests for the behavior we expect for convertToURI. Would probably be easy to break this out into a separate PR?

Comment (malhotrashivam) [reply Base/src/main/java/io/deephaven/base/FileUtils.java:295]: So the new parquet S3 unit tests do test this code path, but you are right, I can add some dedicated tests for convertToURI and move it to a separate PR.

File: extensions/parquet/table/src/test/java/io/deephaven/parquet/table/ParquetS3SimpleMinIOTest.java (hunk 6 above)
Comment (devinrsmith) [reply extensions/parquet/table/src/test/java/io/deephaven/parquet/table/ParquetS3SimpleMinIOTest.java]: Well, I _think_ it should be `assumeTrue` - unconditionally causing unit test failure for a specific platform is not ideal; the idea is that users who do something like `./gradlew check` on OSX should _not_ fail because of the MinIO test (the tests should essentially be skipped). To more specifically test this out on OSX, ensure that `./gradlew extensions-parquet-table:testOutOfBand` completes successfully.

Comment (devinrsmith) [reply extensions/parquet/table/src/test/java/io/deephaven/parquet/table/ParquetS3SimpleMinIOTest.java]: Actually, I think you are right in using `assumeFalse`.

**Review by devinrsmith — APPROVED** [review]

**Review by rcaudy — APPROVED** [review]
